### DISCUSSION TITLE
port forwarding of `python3 -m http.server`

### DISCUSSION THREAD

**Discussion by dragoncoder047** [community/community#64232] [Codespaces] [answered]
### Select Topic Area

Bug

### Body

So I have a static web app in a github repo. HTML, CSS, JS, etc, and no server side logic. The release version is deployed using Github Pages and works.

Prevoiusly when I was on my home computer I made a little bash script that serves all the files on localhost, and then I set chrome to trust localhost as secure.

```console
me@computer:~ $ cat serve
#! /bin/bash
sudo python3 -m http.server -b localhost 80
me@computer:~ $ ./serve
Serving HTTP on ::1 port 80 (http://[::1]:80/) ...
```

and then I open http://localhost/ and I am able to test the dev version of the app.

Now in codespaces when I run that `./serve` it says "Your application running on port 80 is available," I then go to the ports tab and change it to HTTPS and public, **but when I open the forwarded URL all I get is 404 or 401.**

what is causing it to break?

**Comment by Jaber-Saed** [comment]
In GitHub Codespaces, the behavior might differ from what you experienced on your local machine due to various factors, including networking configurations, security policies, and Codespaces' own environment setup.

Here are a few things to consider and troubleshoot:

1. **Port Binding**: It's possible that port 80 is already in use within the Codespaces environment, leading to conflicts. You might want to try using a different port to serve your application.

2. **Firewall or Networking Restrictions**: Codespaces might be running in a restricted environment, which could be causing issues with accessing port 80 or other networking-related activities. GitHub Codespaces runs in a cloud environment, and certain network configurations might not be fully compatible with local development setups.

3. **HTTPS and SSL/TLS**: When you switch to HTTPS in the ports tab, your application should be served over HTTPS, which requires an SSL/TLS certificate. If your application doesn't have proper SSL/TLS configuration, it might lead to 404 or 401 errors. You would need to set up HTTPS for your application by configuring SSL/TLS certificates or using tools like `mkcert` to generate self-signed certificates for development.

4. **Base URL Configuration**: Your web app might have certain base URLs or paths set up that are not working as expected when accessed over a different domain or protocol (HTTP vs. HTTPS). Check your application's HTML, CSS, and JS files to make sure they are referencing resources correctly.

5. **GitHub Pages Compatibility**: While your app works fine on GitHub Pages, GitHub Codespaces might have different restrictions or configurations. Ensure that your app's dependencies and configurations are compatible with Codespaces.

6. **Authentication or Access Control**: The 401 error could be related to authentication or access control settings within your app. Make sure your app is not relying on any specific authentication mechanisms that are not set up in the Codespaces environment.

7. **Debugging and Logs**: Try to access logs and error messages to get more insights into what's going wrong. You can open the developer console in your browser to see if there are any error messages related to resource loading or security issues.

To get more specific help with troubleshooting your GitHub Codespaces setup, it might be helpful to consult GitHub's documentation or seek assistance from their support channels. Additionally, if your goal is to create a consistent development environment, you might consider using Docker or other containerization solutions to ensure that the environment matches across different setups.

**Reply by dragoncoder047** [reply]
> * **HTTPS and SSL/TLS**: When you switch to HTTPS in the ports tab, your application should be served over HTTPS, which requires an SSL/TLS certificate. If your application doesn't have proper SSL/TLS configuration, it might lead to 404 or 401 errors. You would need to set up HTTPS for your application by configuring SSL/TLS certificates or using tools like `mkcert` to generate self-signed certificates for development.

D'oh! I thought that if you selected HTTPS it automatically generated a certificate for you. Going to have to try Flask and port 4443 and see if that works. Will update.

**Reply by dragoncoder047** [reply]
Also, [this](https://github.com/dragoncoder047/langton-music/) is the app in question.

**Comment by dragoncoder047** [comment, marked as answer]
Flask worked.

```console
codespace@codespaces: /workspaces/langton-music $ cat serve
#! /bin/bash
if [[ "$USER" -eq codespace ]]
then
echo "Using Flask to serve http://localhost:4443/ -> port forwarding"
sudo python3 <<PYTHON
import flask, os
app = flask.Flask("test")
@app.route("/")
def main():return flask.redirect("/index.html")
@app.route("/<path:file>")
def files(file):return flask.send_from_directory(os.getcwd(), file, max_age=0)
app.run("localhost", 4443, ssl_context="adhoc")
PYTHON
else
sudo python3 -m http.server -b localhost 80
fi
codespace@codespaces: /workspaces/langton-music $ ./serve
Using Flask to serve http://localhost:4443/ -> port forwarding
 * Serving Flask app 'test'
 * Debug mode: off
WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on https://localhost:4443
Press CTRL+C to quit
```

for future reference: don't forget to `sudo pip3 install cryptography Flask` first
